datanode/snapshot/service_create_snapshot.go

```diff
@@ -0,0 +1,257 @@
+	"code.vegaprotocol.io/vega/logging"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+type Meta struct {
+	ChainID                  string
+	HeightFrom               int64
+	HeightTo                 int64
+	CurrentStateSnapshotFile string
+	HistorySnapshotFile      string
+	WriteLockFile            string
+	DatabaseVersion          int64
+}
+
+func (m Meta) String() string {
+	return fmt.Sprintf("{Chain Id:%s, From Height:%d, To Height:%d, Current State Snapshot File:%s, History Snapshot File:%s}",
+		m.ChainID, m.HeightFrom, m.HeightTo, m.CurrentStateSnapshotFile, m.HistorySnapshotFile)
+}
+
+func (b *Service) CreateSnapshot(ctx context.Context, chainID string, fromHeight int64, toHeight int64) (Meta, error) {
+	var err error
+
+	if len(chainID) == 0 {
+		return Meta{}, fmt.Errorf("chain id is required")
+	}
+
+	start := time.Now()
+	b.log.Infof("creating snapshot for chain %s, from height %d, to height %d", chainID, fromHeight, toHeight)
+
+	conn, err := pgxpool.Connect(ctx, b.connConfig.GetConnectionString())
+	if err != nil {
+		return Meta{}, fmt.Errorf("unable to connect to database: %w", err)
+	}
+	defer conn.Close()
+
+	dbMetaData, err := NewDatabaseMetaData(ctx, b.connConfig)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to get data dump metadata: %w", err)
+	}
+
+	currentSnapshot := NewCurrentSnapshot(chainID, toHeight)
+	historySnapshot := NewHistorySnapshot(chainID, fromHeight, toHeight)
+
+	meta := Meta{
+		ChainID:                  chainID,
+		HeightFrom:               fromHeight,
+		HeightTo:                 toHeight,
+		CurrentStateSnapshotFile: filepath.Join(b.snapshotsPath, currentSnapshot.CompressedFileName()),
+		HistorySnapshotFile:      filepath.Join(b.snapshotsPath, historySnapshot.CompressedFileName()),
+		WriteLockFile:            filepath.Join(b.snapshotsPath, WriteLockFileName(chainID, toHeight)),
+		DatabaseVersion:          dbMetaData.DatabaseVersion,
+	}
+
+	currentDataDir, err := b.createUncompressedDataDirectory(currentSnapshot.UncompressedDataDir())
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to create uncompressed data directory for current snapshot:%w", err)
+	}
+
+	historyDataDir, err := b.createUncompressedDataDirectory(historySnapshot.UncompressedDataDir())
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to create uncompressed data directory for history snapshot:%w", err)
+	}
+
+	copyDataTx, err := conn.Begin(ctx)
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to begin copy table data transaction: %w", err)
+	}
+
+	_, err = copyDataTx.Exec(ctx, "SET TIME ZONE 'UTC'")
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to set timezone to UTC:%w", err)
+	}
+
+	_, err = copyDataTx.Exec(ctx, "SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+	if err != nil {
+		return Meta{}, fmt.Errorf("failed to set transaction isolation level to serilizable:%w", err)
+	}
+
+	writeLockFileName := filepath.Join(b.snapshotsPath, WriteLockFileName(chainID, toHeight))
```

Comment:
I used to use that in the path to create lock files. https://github.com/gofrs/flock

I realise it's really easy to do yourself, but using that help abstract away the details / filesystem stuff in the middle of the logic here and there.

Reply:
I think the semantics are slightly different, in this case the lock file is temporary thing exists while the data is being created, once its removed it basically means the write is complete and the data is fully dumped, after that point it is removed. In the case of flock it states:
`This will release the lock and close the underlying file descriptor.
// It will not remove the file from disk, that's up to your application.`

Which sort of defeats the point of using it.

Reply:
I think `flock` is what we want here because it also locks the file from edits and more importantly being deleted by any other process while locked. The fact `flock` leaves the the file in existence afterwards is irrelevant because its not the existence of the file that means its "locked".

I'm guessing this lock just exists to prevent listing an "in progress" snapshot when you if you call `get-all-available-history` when a snapshot is still in progress? I'm wondering if we should just cut the complexity and have a single file lock for "data node is still running" because we'll need something to prevent someone calling `load-all-available-history` if the data-node is running at all, regardless of is there is a snapshot in progress.

Reply:
regarding the load all available, that ensures the datanode is dead before proceeding, it kills all connections to the database before proceeding to prevent accidental concurrent running.